consensus/src/experimental/buffer_manager.rs

```diff
@@ -254,6 +184,112 @@ impl StateManager {
         Ok(())
     }
 
+    /// this function updates the buffer according to the response from the execution phase
+    /// it also initiates a request to the signing phase.
+    async fn update_buffer_execution_resp(
+        &mut self,
+        execution_resp: ExecutionResponse,
+    ) -> anyhow::Result<()> {
+        let ExecutionResponse { inner } = execution_resp;
+
+        let executed_blocks = inner?;
+
+        let current_cursor = get_next(&self.execution_root); // must be Some
+
+        if current_cursor.is_some() {
+            // update buffer
+            let buffer_item = take_elem(&current_cursor);
+            if let BufferItem::Ordered(ordered_box) = &buffer_item {
+                if ordered_box.ordered_blocks.first().unwrap().id()
+                    != executed_blocks.first().unwrap().id()
+                {
+                    // an sync req happened before the response
+                    // we do nothing except putting the item back
+                    // the process_execution_resp function will retry the next ordered batch
+                    set_elem(&current_cursor, buffer_item);
+                    return Ok(());
+                }
+                for (ordered_block, executed_block) in
+                    zip(ordered_box.ordered_blocks.iter(), executed_blocks.iter())
+                {
+                    assert_eq!(ordered_block.id(), executed_block.id())
+                }
+
+                // push eto the signing phase
+                let commit_ledger_info = LedgerInfo::new(
+                    executed_blocks.last().unwrap().block_info(),
+                    ordered_box
+                        .ordered_proof
+                        .ledger_info()
+                        .consensus_data_hash(),
+                );
+
+                self.signing_phase_tx
+                    .send(SigningRequest {
+                        ordered_ledger_info: ordered_box.ordered_proof.clone(),
+                        commit_ledger_info,
+                    })
+                    .await?;
+
+                set_elem(
+                    &current_cursor,
+                    buffer_item.advance_to_executed(executed_blocks),
+                );
+                self.execution_root = current_cursor;
+            } else {
+                // even if there is a sync happened before the response
+                // the buffer item right after execution root should be an ordered buffer item
+                return Err(anyhow!("Inconsistent buffer item state"));
```

Comment:
isn't this an invariant that should assert?

Reply:
resolved